<a href="https://colab.research.google.com/github/HayeonKimm/project/blob/master/project/python_machine_pjt/toy_machine_pjt/%EB%B6%93%EA%BD%83_%ED%92%88%EC%A2%85_%EC%98%88%EC%B8%A1%ED%95%B4%EB%B3%B4%EC%9E%901021.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 파이썬 머신러닝 완벽가이드 (위키북스) 기반으로 실습.

## 붖꽃 품종 예측 (1021)

## 붓꽃 데이터 세트로 분류를 예측한 프로세스

1.데이터 세트 분리  
2.모델 학습  
3.예측 수행  
4.평가  

In [1]:
from sklearn.datasets import load_iris # 사이킷런에서 자체적으로 제공하는 데이터 세트를 생성하는 모듈의 모임.
from sklearn.tree import DecisionTreeClassifier # 트리기반 ML알고리즘을 구현한 클래스의 모임 
from sklearn.model_selection import train_test_split # 데이터 분리, 최적의 하이퍼 파라미터로 평가하기 위한 다양한 모듈의 모임

In [2]:
import pandas as pd

In [3]:
iris = load_iris() # 원래 자체적으로 주어지는 데이터 

In [4]:
iris_data=iris.data # iris의 데이터만 따로 할당

In [5]:
iris_label = iris.target # 타겟값은 따로 할당.

print('iris target값:', iris_label) # 타겟 값을 보여줘.
print('iris target값 네임:', iris.target_names) # 타겟 이름을 알려줘 

iris target값: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]
iris target값 네임: ['setosa' 'versicolor' 'virginica']


In [6]:
dir(iris) # 대충 이런 것들을 쓸 수 있다.

['DESCR', 'data', 'feature_names', 'filename', 'target', 'target_names']

In [7]:
X_train, X_test, y_train, y_test = train_test_split(iris_data, iris_label, test_size=0.2, random_state = 11)  # 트레이닝 셋 , 테스트 셋으로 나누었다. 테스트 셋 사이즈는 0.2 , 고로 트레이닝 사이즈는 0.8이다. 

# random_state = 11 로 고정시켰다. 수행 할 때마다 동일한 데이터 세트로 분리하기 위해서.

## 학습 데이터를 확보했으니, 머신러닝 분류 알고리즘의 하나인 의사 결정 트리를 이용해 학습과 예측을 수행 해보자.

In [8]:
# DecisionTreeClassifier 객체 생성
dt_clf = DecisionTreeClassifier(random_state=11)

# 학습을 수행하자 !
dt_clf.fit(X_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=11, splitter='best')

## 학습이 완료됐다.

In [9]:
# 학습이 완료 됐다. 테스트 데이터로 잘 수행되었는지 알아보자.

pred = dt_clf.predict(X_test) # X 테스트 데이터로 y 테스트 데이터 예측을 수행해보자. 

## 일반적으로 머신러닝 모델의 성능 평가 방법은 여러 가지가 있지만, 여기서는 '정확도' 를 이용해 측정하겠다.

In [10]:
# accuracy_score 임포트
from sklearn.metrics import accuracy_score 

print('예측 정확도: {0:.4f}'.format(accuracy_score(y_test,pred)))

예측 정확도: 0.9333


##예측 정확도는 0.9333이 나왔다.
--------------------------

## 번외로 트레이닝 , 테스트 셋 비율을 바꿔서 다시 실행해보자.

In [11]:
# 테스트 사이즈 0.1 높여서 0.3으로 실행

X_train, X_test, y_train, y_test = train_test_split(iris_data, iris_label, test_size=0.3, random_state = 12)

# DecisionTreeClassifier 객체 생성
dt_clf = DecisionTreeClassifier(random_state=12)

# 학습을 수행하자 !
dt_clf.fit(X_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=12, splitter='best')

In [12]:
pred = dt_clf.predict(X_test)

from sklearn.metrics import accuracy_score 

print('예측 정확도: {0:.4f}'.format(accuracy_score(y_test,pred)))

예측 정확도: 0.9556


In [13]:
# 테스트 사이즈 0.1 낮춰서 0.1으로 실행


X_train, X_test, y_train, y_test = train_test_split(iris_data, iris_label, test_size=0.1, random_state = 13)

# DecisionTreeClassifier 객체 생성
dt_clf = DecisionTreeClassifier(random_state=13)

# 학습을 수행하자 !
dt_clf.fit(X_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=13, splitter='best')

In [14]:
pred = dt_clf.predict(X_test)

from sklearn.metrics import accuracy_score 

print('예측 정확도: {0:.4f}'.format(accuracy_score(y_test,pred)))

예측 정확도: 0.8667


# 결론 : 테스트 사이즈 0.3 이 가장 정확도가 높다 ! 